v2/tracing/tracing.go

```diff
@@ -5,57 +5,80 @@ import (
-		ctx, span := tab.StartSpanWithRemoteParent(ctx, "go-shuttle.receiver.Handle", carrierAdapter(message))
+		ctx, span := applyTracingMiddleWare(ctx, message)
 		defer span.End()
-
-		if message != nil {
-			span.AddAttributes(tab.StringAttribute("message.id", message.MessageID))
-			if message.CorrelationID != nil {
-				span.AddAttributes(tab.StringAttribute("message.correlationId", *message.CorrelationID))
-			}
-			if message.ScheduledEnqueueTime != nil {
-				span.AddAttributes(tab.StringAttribute("message.scheduledEnqueuedTime", message.ScheduledEnqueueTime.String()))
-			}
-			if message.TimeToLive != nil {
-				span.AddAttributes(tab.StringAttribute("message.ttl", message.TimeToLive.String()))
-			}
-		} else {
-			span.Logger().Info("warning: message is nil")
-		}
 		handler(ctx, settler, message)
 	}
 }
 
-// carrierAdapter wraps a Received Message so that it implements the tab.Carrier interface
-func carrierAdapter(message *azservicebus.ReceivedMessage) tab.Carrier {
-	return &messageWrapper{message: message}
-}
+func applyTracingMiddleWare(ctx context.Context, message *azservicebus.ReceivedMessage) (context.Context, trace.Span) {
+	var span trace.Span
+	var attrs []attribute.KeyValue
+	var traceCarrier map[string]string
 
-type messageWrapper struct {
-	message *azservicebus.ReceivedMessage
-}
+	if message != nil {
+		attrs = append(attrs, attribute.String("message.id", message.MessageID))
 
-// Set implements tab.Carrier interface
-func (mw *messageWrapper) Set(key string, value interface{}) {
-	if mw.message != nil {
-		if mw.message.ApplicationProperties == nil {
-			mw.message.ApplicationProperties = make(map[string]interface{})
+		if message.CorrelationID != nil {
+			attrs = append(attrs, attribute.String("message.correlationId", *message.CorrelationID))
 		}
-		mw.message.ApplicationProperties[key] = value
+
+		if message.ScheduledEnqueueTime != nil {
+			attrs = append(attrs, attribute.String("message.scheduledEnqueuedTime", message.ScheduledEnqueueTime.String()))
+		}
+
+		if message.TimeToLive != nil {
+			attrs = append(attrs, attribute.String("message.ttl", message.TimeToLive.String()))
+		}
+
+		if message.ApplicationProperties != nil {
+			traceCarrier = message.ApplicationProperties[traceCarrierField].(map[string]string)
+		}
+	}
+
+	// extract the remote trace context from the traceCarrier if exists,
+	if traceCarrier != nil {
+		ctx, _ = getRemoteParentSpan(ctx, traceCarrier)
 	}
+
+	ctx, span = tracer.Start(ctx, "receiver.Handle")
+	span.SetAttributes(attrs...)
+	return ctx, span
 }
 
-// GetKeyValues implements tab.Carrier interface
-func (mw *messageWrapper) GetKeyValues() map[string]interface{} {
-	if mw.message != nil {
-		return mw.message.ApplicationProperties
+// This method will extract the remote trace context from the traceCarrier if exists,
+// and inject it into the passed in context. If the traceCarrier doesn't contain a
+// valid trace context, the passed in context will be returned as is.
```

Comment:
are you sure this is the right behavior?
I was expecting that a new traceContext would be added to the span? but that may be wrong.

Reply:
From the otel implementation, you are right:

// The returned Context will be a copy of ctx and contain the extracted
// tracecontext as the remote SpanContext. If the extracted tracecontext is
// invalid, the passed ctx will be returned directly instead.

resolving